File: LinkedList/problemLinkedList.cpp
#include "LinkedList.h"

/**
 *  
 * problem related to linked list. 
 * 
*/

LinkedListNode* kthNodeFromLast(LinkedList *list, int k); // get the kth from last. 

LinkedListNode* kthNodeFromLast(LinkedListNode *node, int &i, int k); // get the kth node from last. 

int main(int argc, char* argv[]) {

    LinkedList *list = new LinkedList(); 
    for(int i = 0; i < 20; i++) {
        list->insertKey(rand() % 100); 
    }
    list->printList(); 
    cout << endl; 
    int k = 0; 
    cin >> k; 
    cout << kthNodeFromLast(list, k)->getData() << endl; 

}


LinkedListNode* kthNodeFromLast(LinkedList *list, int k) {
    int i = 0; 
    return kthNodeFromLast(list->getHead(), i, k);  
}

LinkedListNode* kthNodeFromLast(LinkedListNode *head, int &i, int k) {
    if(head == NULL) {
        return NULL; 
    } else {
        LinkedListNode *node = kthNodeFromLast(head->getNext(), i, k); 
        i += 1; 
        if(i == k) {
            return head; 
        } 
        return node; 
    }
}

File: LinkedList/problemLinkedList.py
# filename : problemLinkedList.py 

# problem 1 : remove duplicates. 

import LinkedList 


def removeDuplicate(linkedlist): 
    current = linkedlist.head 
    previous = current 
    aux_data = []
    while current is not None:
        if current.data in aux_data: 
            previous.next = current.next 
        else: 
            aux_data.append(current.data) # append data 
            previous = previous.next 
        current = current.next       
      
def makeLinkedList(lst): 
    linkedlist = LinkedList.LinkedList()
    for ele in lst: 
        linkedlist.insertKey(ele) 
    return linkedlist 

if __name__ == '__main__': 
    data = [12, 12, 44, 1, 131, 11, 1, 32, 1] # list 
    linkedlist = makeLinkedList(data) # get linked list. 
    linkedlist.printList() # print linked list. 
    removeDuplicate(linkedlist)
    print() # get an empty line. 
    linkedlist.printList()   